File: src/big_config/selmer_filters.clj
(ns big-config.selmer-filters
  (:require [clojure.string :as str]
            [selmer.parser :as selmer]))

(defn normalize-delimiters
  ([] (normalize-delimiters {}))
  ([delimiters]
   {:tag-open (or (:tag-open delimiters) (:tagOpen delimiters) "{")
    :tag-close (or (:tag-close delimiters) (:tagClose delimiters) "}")
    :filter-open (or (:filter-open delimiters) (:filterOpen delimiters) "{")
    :filter-close (or (:filter-close delimiters) (:filterClose delimiters) "}")
    :tag-second (or (:tag-second delimiters) (:tagSecond delimiters) "%")
    :short-comment-second (or (:short-comment-second delimiters) (:shortCommentSecond delimiters) "#")}))

(defn whitespace-control
  ([input] (whitespace-control input {}))
  ([input delimiters]
   (let [d (normalize-delimiters delimiters)
         opening-tags #{(str (:tag-open d) (:filter-open d) "-")
                        (str (:tag-open d) (:tag-second d) "-")}
         closing-tags #{(str "-" (:tag-close d) (:filter-close d))
                        (str "-" (:tag-second d) (:tag-close d))}]
     (loop [output ""
            rest (seq (str input))
            tag ""]
       (if-not rest
         output
         (let [x (str (first rest))
               tag' (str tag x)
               tag' (if (= 4 (count tag')) (subs tag' 1) tag')]
           (cond
             (contains? opening-tags tag')
             (recur (str (str/trimr (subs output 0 (max 0 (- (count output) 2))))
                         (subs tag' 0 2))
                    (next rest)
                    tag')

             (contains? closing-tags tag')
             (recur (str (subs output 0 (max 0 (- (count output) 2)))
                         (subs tag' 1 3))
                    (seq (str/triml (apply str (next rest))))
                    tag')

             :else
             (recur (str output x) (next rest) tag'))))))))

(selmer/add-filter! "lookup-env" (fn [x] (System/getenv (str x))))
(selmer/add-filter! "->file" (fn [n] (-> (str n) (str/replace "." "/") (str/replace "-" "_"))))
(selmer/add-filter! "remove-https" (fn [url] (str/replace (str url) #"^https://" "")))

File: src/big_config/selmer_filters.py
from __future__ import annotations

import os
from typing import Any

from selmer import add_filter


def lookup_env(x: Any) -> str | None:
    return os.environ.get(str(x))


def to_file(n: Any) -> str:
    return str(n).replace(".", "/").replace("-", "_")


add_filter("lookup-env", lookup_env)
add_filter("->file", to_file)


def _delim(delimiters: dict[str, Any], key: str, default: str) -> str:
    value = delimiters.get(key, delimiters.get(key.replace("-", "_"), default))
    return str(value)


def whitespace_control(s: str, delimiters: dict[str, Any] | None = None) -> str:
    """Implement Selmer's ``{{-`` / ``-}}`` whitespace trim workaround."""

    delimiters = delimiters or {}
    d = {
        "tag-open": _delim(delimiters, "tag-open", "{"),
        "tag-close": _delim(delimiters, "tag-close", "}"),
        "filter-open": _delim(delimiters, "filter-open", "{"),
        "filter-close": _delim(delimiters, "filter-close", "}"),
        "tag-second": _delim(delimiters, "tag-second", "%"),
    }
    opening_tags = {
        d["tag-open"] + d["filter-open"] + "-",
        d["tag-open"] + d["tag-second"] + "-",
    }
    closing_tags = {
        "-" + d["tag-close"] + d["filter-close"],
        "-" + d["tag-second"] + d["tag-close"],
    }
    output = ""
    tag = ""
    i = 0
    while i < len(s):
        x = s[i]
        tag = (tag + x)[-3:]
        if tag in opening_tags:
            output = output[:-2].rstrip() + tag[:2]
        elif tag in closing_tags:
            output = output[:-2] + tag[1:3]
            i += 1
            while i < len(s) and s[i].isspace():
                i += 1
            continue
        else:
            output += x
        i += 1
    return output
